cpp/src/arrow/dataset/discovery.cc

```diff
@@ -28,43 +28,74 @@
 #include "arrow/dataset/partition.h"
 #include "arrow/dataset/type_fwd.h"
 #include "arrow/filesystem/path_tree.h"
+#include "arrow/filesystem/path_util.h"
 #include "arrow/status.h"
 
 namespace arrow {
 namespace dataset {
 
 FileSystemDataSourceDiscovery::FileSystemDataSourceDiscovery(
-    fs::FileSystem* filesystem, std::string base_dir, std::vector<fs::FileStats> files,
-    std::shared_ptr<FileFormat> format)
+    fs::FileSystem* filesystem, std::vector<fs::FileStats> files,
+    std::shared_ptr<FileFormat> format, FileSystemDiscoveryOptions options)
     : fs_(filesystem),
-      base_dir_(std::move(base_dir)),
       files_(std::move(files)),
-      format_(std::move(format)) {}
+      format_(std::move(format)),
+      options_(std::move(options)) {}
+
+std::vector<fs::FileStats> FilterFilesByPrefix(const std::vector<std::string>& prefixes,
+                                               std::vector<fs::FileStats> files) {
+  std::vector<fs::FileStats> filtered;
+  auto matches_prefix = [&prefixes](const fs::FileStats& file) -> bool {
+    // Ignore non-file.
+    if (!file.IsFile()) return true;
+
+    auto dir_base = fs::internal::GetAbstractPathParent(file.path());
+    util::string_view basename{dir_base.second};
+
+    auto has_prefix = [&basename](const std::string& prefix) -> bool {
+      return !prefix.empty() && basename.starts_with(prefix);
+    };
+
+    return !std::any_of(prefixes.cbegin(), prefixes.cend(), has_prefix);
+  };
+
+  for (const auto& stat : files) {
+    if (matches_prefix(stat)) {
+      filtered.push_back(stat);
+    }
+  }
 
-Status FileSystemDataSourceDiscovery::Make(fs::FileSystem* filesystem,
-                                           std::string base_dir,
-                                           std::vector<fs::FileStats> files,
-                                           std::shared_ptr<FileFormat> format,
-                                           std::shared_ptr<DataSourceDiscovery>* out) {
-  out->reset(
-      new FileSystemDataSourceDiscovery(filesystem, std::move(base_dir), files, format));
-  return Status::OK();
+  return filtered;
 }
 
 Status FileSystemDataSourceDiscovery::Make(fs::FileSystem* filesystem,
                                            std::vector<fs::FileStats> files,
                                            std::shared_ptr<FileFormat> format,
+                                           FileSystemDiscoveryOptions options,
                                            std::shared_ptr<DataSourceDiscovery>* out) {
-  return Make(filesystem, "", std::move(files), std::move(format), out);
+  const auto& prefixes = options.ignore_prefixes;
+  std::vector<fs::FileStats> filtered =
+      prefixes.empty() ? std::move(files) : FilterFilesByPrefix(prefixes, files);
+
+  out->reset(new FileSystemDataSourceDiscovery(filesystem, std::move(filtered),
+                                               std::move(format), std::move(options)));
```

Comment:
The doccomment for this method indicates that ignored files will be part of the resulting datasource, but wouldn't that require that you also include the files which match ignored prefixes in this construction somehow? I think the best thing for now would be to amend the doccomment to state that files matching any ignored prefix will be dropped.

Reply:
Which one?

Reply:
https://github.com/apache/arrow/pull/5811/files#diff-91c3d3ad9e9a52dd2886621a692bcf6eR91-R92 "The ignored files will still be part of the DataSource, but will not have partition information."

Reply:
> For the purpose of applying the partition scheme

Reply:
Do you want me to clarify the comment? This is now part of options and only scoped to the partition_base_dir.